### PR TITLE
Fixes issue #1647

### PR DESCRIPTION
**Pull request by Smokexyz**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Allow skills and attacks with flags `BCT_ENEMY` and `BCT_PARTY` to affect "guild enemy units" within the party on maps with flags `gvg` and `gvg_noparty`.

**Affected Branches:** master

[//]: # (Master? Slave?)

**Issues addressed:** #1647 

[//]: # (Issue Tracker Number if any.)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

---
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1648)

### CODE REVIEW

File: src/map/battle.c [hunk 1]
```diff
@@ -6814,14 +6814,20 @@ int battle_check_target( struct block_list *src, struct block_list *target,int f
...
 			int s_party = status->get_party_id(s_bl);
 			int s_guild = status->get_guild_id(s_bl);
-
-			if( s_party && s_party == status->get_party_id(t_bl)
-			 && !(map->list[m].flag.pvp && map->list[m].flag.pvp_noparty)
-			 && !(map_flag_gvg(m) && map->list[m].flag.gvg_noparty && !( s_guild && s_guild == status->get_guild_id(t_bl) ))
-			 && (!map->list[m].flag.battleground || sbg_id == tbg_id) )
+			int t_guild = status->get_guild_id(t_bl);
+			
+			if (map_flag_gvg(m) && map->list[m].flag.gvg_noparty && !(s_guild && s_guild == t_guild)) {
```
Comment (MishimaHaruna) [inline src/map/battle.c]: I believe this should still be inside a same-party check. We don't want allied guildsmen outside the party to be considered party members

Comment (Smokexyz) [inline src/map/battle.c]: Oops, fixed.

File: src/map/battle.c [hunk 2]
```diff
@@ -6814,14 +6814,21 @@ int battle_check_target( struct block_list *src, struct block_list *target,int f
...
+			int t_guild = status->get_guild_id(t_bl);
+			
+			if (s_party && s_party == status->get_party_id(t_bl)) {
+				if (map_flag_gvg(m) && map->list[m].flag.gvg_noparty && !(s_guild && s_guild == t_guild)) {
+					if (t_guild && guild->isallied(s_guild, t_guild))
+						state |= BCT_PARTY;
+					else
+						state |= flag&BCT_ENEMY ? BCT_ENEMY : BCT_PARTY;
+				} else if (!(map->list[m].flag.pvp && map->list[m].flag.pvp_noparty)
+					&& (!map->list[m].flag.battleground || sbg_id == tbg_id)) {
```
Comment (MishimaHaruna) [inline src/map/battle.c]: Should this also have a final `else` branch that sets BCT_ENEMY if this condition isn't met? (example: in battlegrounds when the `sbg_id` and `tbg_id` don't match, or in pvp_noparty)

Comment (Smokexyz) [inline src/map/battle.c]: Damn, yup.

Comment (Smokexyz) [inline src/map/battle.c]: Fixed.

**Comment by MishimaHaruna** [comment]
@skyleo can you confirm that with this PR applied, the behavior is correct wrt your issue #1647 ?

**Comment by skyleo** [comment]
@MishimaHaruna Yes, I confirm that the described behavior is correct with this change.
I checked a few other things to make sure no annoying side-effects are there, couldn't find anything.
We'll still have to keep an eye open for that though the comming days imo.

**Comment by MishimaHaruna** [comment]
Ok, I'll be merging this, thank you

**Pull request merged**